File: frontend/src/types/resident-weekly-requirement.ts
/**
 * Resident Weekly Requirement Types
 *
 * Types for managing weekly scheduling requirements per rotation template.
 * Controls FM clinic and specialty session frequencies, academic requirements,
 * and protected time slots for resident schedules.
 */

// ============================================================================
// Core Types
// ============================================================================

/**
 * Days of the week as numbers (0 = Sunday, 6 = Saturday)
 */
export type DayOfWeek = 0 | 1 | 2 | 3 | 4 | 5 | 6;

/**
 * Time slot identifiers for protected slots
 */
export type TimeSlot =
  | 'mon_am' | 'mon_pm'
  | 'tue_am' | 'tue_pm'
  | 'wed_am' | 'wed_pm'
  | 'thu_am' | 'thu_pm'
  | 'fri_am' | 'fri_pm'
  | 'sat_am' | 'sat_pm'
  | 'sun_am' | 'sun_pm';

/**
 * Protected slot activity types
 */
export type ProtectedSlotType =
  | 'conference'
  | 'academics'
  | 'didactics'
  | 'research'
  | 'admin'
  | 'wellness';

/**
 * Record of protected time slots mapped to their activity type
 */
export type ProtectedSlots = Partial<Record<TimeSlot, ProtectedSlotType>>;

/**
 * Resident Weekly Requirement - defines scheduling constraints for a rotation
 */
export interface ResidentWeeklyRequirement {
  /** Unique identifier */
  id: string;
  /** Associated rotation template ID */
  rotationTemplateId: string;
  /** Minimum FM clinic half-days per week (0-14) */
  fmClinicMinPerWeek: number;
  /** Maximum FM clinic half-days per week (0-14) */
  fmClinicMaxPerWeek: number;
  /** Minimum specialty half-days per week */
  specialtyMinPerWeek: number;
  /** Maximum specialty half-days per week */
  specialtyMaxPerWeek: number;
  /** Whether Wednesday AM academics are required */
  academicsRequired: boolean;
  /** Protected time slots mapped to activity type */
  protectedSlots: ProtectedSlots;
  /** Days of week when clinic sessions are allowed (0=Sun, 1=Mon, ..., 6=Sat) */
  allowedClinicDays: DayOfWeek[];
  /** Creation timestamp */
  createdAt?: string;
  /** Last update timestamp */
  updatedAt?: string;
}

// ============================================================================
// API Request/Response Types
// ============================================================================

/**
 * Request to create a new weekly requirement
 */
export interface ResidentWeeklyRequirementCreate {
  rotationTemplateId: string;
  fmClinicMinPerWeek: number;
  fmClinicMaxPerWeek: number;
  specialtyMinPerWeek: number;
  specialtyMaxPerWeek: number;
  academicsRequired?: boolean;
  protectedSlots?: ProtectedSlots;
  allowedClinicDays?: DayOfWeek[];
}

/**
 * Request to update an existing weekly requirement
 */
export interface ResidentWeeklyRequirementUpdate {
  fmClinicMinPerWeek?: number;
  fmClinicMaxPerWeek?: number;
  specialtyMinPerWeek?: number;
  specialtyMaxPerWeek?: number;
  academicsRequired?: boolean;
  protectedSlots?: ProtectedSlots;
  allowedClinicDays?: DayOfWeek[];
}

// ============================================================================
// Form Types
// ============================================================================

/**
 * Form state for editing weekly requirements
 */
export interface WeeklyRequirementFormData {
  fmClinicMinPerWeek: number;
  fmClinicMaxPerWeek: number;
  specialtyMinPerWeek: number;
  specialtyMaxPerWeek: number;
  academicsRequired: boolean;
  protectedSlots: ProtectedSlots;
  allowedClinicDays: DayOfWeek[];
}

/**
 * Validation errors for the form
 */
export interface WeeklyRequirementFormErrors {
  fmClinic?: string;
  specialty?: string;
  general?: string;
}

// ============================================================================
// Display Constants
// ============================================================================

/**
 * Day of week labels for UI display
 */
export const DAY_OF_WEEK_LABELS: Record<DayOfWeek, string> = {
  0: 'Sun',
  1: 'Mon',
  2: 'Tue',
  3: 'Wed',
  4: 'Thu',
  5: 'Fri',
  6: 'Sat',
};

/**
 * Full day of week labels
 */
export const DAY_OF_WEEK_FULL_LABELS: Record<DayOfWeek, string> = {
  0: 'Sunday',
  1: 'Monday',
  2: 'Tuesday',
  3: 'Wednesday',
  4: 'Thursday',
  5: 'Friday',
  6: 'Saturday',
};

/**
 * Time slot display labels
 */
export const TIME_SLOT_LABELS: Record<TimeSlot, string> = {
  monAm: 'Mon AM',
  monPm: 'Mon PM',
  tueAm: 'Tue AM',
  tuePm: 'Tue PM',
  wedAm: 'Wed AM',
  wedPm: 'Wed PM',
  thuAm: 'Thu AM',
  thuPm: 'Thu PM',
  friAm: 'Fri AM',
  friPm: 'Fri PM',
  satAm: 'Sat AM',
  satPm: 'Sat PM',
  sunAm: 'Sun AM',
  sunPm: 'Sun PM',
};

/**
 * Protected slot type labels and colors
 */
export const PROTECTED_SLOT_TYPE_CONFIG: Record<
  ProtectedSlotType,
  { label: string; color: string; bgColor: string }
> = {
  conference: {
    label: 'Conference',
    color: 'text-amber-400',
    bgColor: 'bg-amber-500/20',
  },
  academics: {
    label: 'Academics',
    color: 'text-blue-400',
    bgColor: 'bg-blue-500/20',
  },
  didactics: {
    label: 'Didactics',
    color: 'text-cyan-400',
    bgColor: 'bg-cyan-500/20',
  },
  research: {
    label: 'Research',
    color: 'text-purple-400',
    bgColor: 'bg-purple-500/20',
  },
  admin: {
    label: 'Admin',
    color: 'text-slate-400',
    bgColor: 'bg-slate-500/20',
  },
  wellness: {
    label: 'Wellness',
    color: 'text-green-400',
    bgColor: 'bg-green-500/20',
  },
};

/**
 * All protected slot types as an array for iteration
 */
export const PROTECTED_SLOT_TYPES: ProtectedSlotType[] = [
  'conference',
  'academics',
  'didactics',
  'research',
  'admin',
  'wellness',
];

/**
 * All time slots in order
 */
export const TIME_SLOTS: TimeSlot[] = [
  'sun_am', 'sun_pm',
  'mon_am', 'mon_pm',
  'tue_am', 'tue_pm',
  'wed_am', 'wed_pm',
  'thu_am', 'thu_pm',
  'fri_am', 'fri_pm',
  'sat_am', 'sat_pm',
];

/**
 * Default values for a new weekly requirement
 */
export const DEFAULT_WEEKLY_REQUIREMENT: WeeklyRequirementFormData = {
  fmClinicMinPerWeek: 0,
  fmClinicMaxPerWeek: 4,
  specialtyMinPerWeek: 0,
  specialtyMaxPerWeek: 4,
  academicsRequired: true,
  protectedSlots: { wedAm: 'conference' },
  allowedClinicDays: [1, 2, 3, 4, 5], // Mon-Fri
};

// ============================================================================
// Validation Helpers
// ============================================================================

/**
 * Validates min <= max constraint
 */
export function validateMinMax(min: number, max: number): boolean {
  return min <= max;
}

/**
 * Validates half-day count is within valid range (0-14)
 */
export function validateHalfDayRange(value: number): boolean {
  return value >= 0 && value <= 14;
}

/**
 * Validates the entire form data
 */
export function validateWeeklyRequirement(
  data: WeeklyRequirementFormData
): WeeklyRequirementFormErrors {
  const errors: WeeklyRequirementFormErrors = {};

  // FM Clinic validation
  if (!validateHalfDayRange(data.fm_clinic_min_per_week) ||
      !validateHalfDayRange(data.fm_clinic_max_per_week)) {
    errors.fm_clinic = 'FM clinic values must be between 0 and 14';
  } else if (!validateMinMax(data.fm_clinic_min_per_week, data.fm_clinic_max_per_week)) {
    errors.fm_clinic = 'FM clinic minimum cannot exceed maximum';
  }

  // Specialty validation
  if (!validateHalfDayRange(data.specialty_min_per_week) ||
      !validateHalfDayRange(data.specialty_max_per_week)) {
    errors.specialty = 'Specialty values must be between 0 and 14';
  } else if (!validateMinMax(data.specialty_min_per_week, data.specialty_max_per_week)) {
    errors.specialty = 'Specialty minimum cannot exceed maximum';
  }

  return errors;
}

/**
 * Checks if form data has validation errors
 */
export function hasValidationErrors(errors: WeeklyRequirementFormErrors): boolean {
  return Object.keys(errors).length > 0;
}

File: backend/app/schemas/resident_weekly_requirement.py
"""Resident weekly requirement schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class ResidentWeeklyRequirementBase(BaseModel):
    """Base schema for resident weekly requirements."""

    fm_clinic_min_per_week: int = Field(
        default=2,
        ge=0,
        le=10,
        description="Minimum FM clinic half-days per week (ACGME: 2 for outpatient)",
    )
    fm_clinic_max_per_week: int = Field(
        default=3, ge=0, le=10, description="Maximum FM clinic half-days per week"
    )
    specialty_min_per_week: int = Field(
        default=0, ge=0, le=10, description="Minimum specialty half-days per week"
    )
    specialty_max_per_week: int = Field(
        default=10, ge=0, le=10, description="Maximum specialty half-days per week"
    )
    academics_required: bool = Field(
        default=True, description="Whether academic time (Wed AM) is required"
    )
    protected_slots: dict[str, str] = Field(
        default_factory=dict,
        description="Protected slots mapping (e.g., {'wed_am': 'conference'})",
    )
    allowed_clinic_days: list[int] = Field(
        default_factory=list,
        description="Allowed clinic days (0=Mon, 4=Fri). Empty = all weekdays allowed",
    )
    specialty_name: str | None = Field(
        default=None, max_length=255, description="Specialty name (e.g., 'Neurology')"
    )
    description: str | None = Field(
        default=None, max_length=1024, description="Optional notes about requirements"
    )

    @field_validator("protected_slots")
    @classmethod
    def validate_protected_slots(cls, v: dict[str, str]) -> dict[str, str]:
        """Validate protected_slots keys are valid slot identifiers."""
        valid_days = ["mon", "tue", "wed", "thu", "fri"]
        valid_times = ["am", "pm"]

        for slot_key in v.keys():
            parts = slot_key.lower().split("_")
            if len(parts) != 2:
                raise ValueError(
                    f"Invalid slot key '{slot_key}'. Format: 'day_time' (e.g., 'wed_am')"
                )
            if parts[0] not in valid_days:
                raise ValueError(
                    f"Invalid day '{parts[0]}' in slot key. Must be one of {valid_days}"
                )
            if parts[1] not in valid_times:
                raise ValueError(
                    f"Invalid time '{parts[1]}' in slot key. Must be 'am' or 'pm'"
                )

        return v

    @field_validator("allowed_clinic_days")
    @classmethod
    def validate_allowed_days(cls, v: list[int]) -> list[int]:
        """Validate allowed_clinic_days are valid weekday integers."""
        for day in v:
            if day < 0 or day > 4:
                raise ValueError(
                    f"Invalid day {day}. Must be 0-4 (Monday=0, Friday=4)"
                )
        return sorted(set(v))  # Remove duplicates and sort

    @model_validator(mode="after")
    def validate_ranges(self) -> "ResidentWeeklyRequirementBase":
        """Validate that min values are less than or equal to max values."""
        if self.fm_clinic_min_per_week > self.fm_clinic_max_per_week:
            raise ValueError(
                f"fm_clinic_min_per_week ({self.fm_clinic_min_per_week}) cannot exceed "
                f"fm_clinic_max_per_week ({self.fm_clinic_max_per_week})"
            )
        if self.specialty_min_per_week > self.specialty_max_per_week:
            raise ValueError(
                f"specialty_min_per_week ({self.specialty_min_per_week}) cannot exceed "
                f"specialty_max_per_week ({self.specialty_max_per_week})"
            )
        return self


class ResidentWeeklyRequirementCreate(ResidentWeeklyRequirementBase):
    """Schema for creating a resident weekly requirement."""

    rotation_template_id: UUID = Field(
        ..., description="UUID of the rotation template this applies to"
    )


class ResidentWeeklyRequirementUpdate(BaseModel):
    """Schema for updating a resident weekly requirement (all fields optional)."""

    fm_clinic_min_per_week: int | None = Field(
        default=None,
        ge=0,
        le=10,
        description="Minimum FM clinic half-days per week",
    )
    fm_clinic_max_per_week: int | None = Field(
        default=None, ge=0, le=10, description="Maximum FM clinic half-days per week"
    )
    specialty_min_per_week: int | None = Field(
        default=None, ge=0, le=10, description="Minimum specialty half-days per week"
    )
    specialty_max_per_week: int | None = Field(
        default=None, ge=0, le=10, description="Maximum specialty half-days per week"
    )
    academics_required: bool | None = Field(
        default=None, description="Whether academic time is required"
    )
    protected_slots: dict[str, str] | None = Field(
        default=None, description="Protected slots mapping"
    )
    allowed_clinic_days: list[int] | None = Field(
        default=None, description="Allowed clinic days"
    )
    specialty_name: str | None = Field(
        default=None, max_length=255, description="Specialty name"
    )
    description: str | None = Field(
        default=None, max_length=1024, description="Optional notes"
    )

    @field_validator("protected_slots")
    @classmethod
    def validate_protected_slots(cls, v: dict[str, str] | None) -> dict[str, str] | None:
        """Validate protected_slots keys if provided."""
        if v is None:
            return v

        valid_days = ["mon", "tue", "wed", "thu", "fri"]
        valid_times = ["am", "pm"]

        for slot_key in v.keys():
            parts = slot_key.lower().split("_")
            if len(parts) != 2:
                raise ValueError(f"Invalid slot key '{slot_key}'")
            if parts[0] not in valid_days or parts[1] not in valid_times:
                raise ValueError(f"Invalid slot key '{slot_key}'")

        return v

    @field_validator("allowed_clinic_days")
    @classmethod
    def validate_allowed_days(cls, v: list[int] | None) -> list[int] | None:
        """Validate allowed_clinic_days if provided."""
        if v is None:
            return v
        for day in v:
            if day < 0 or day > 4:
                raise ValueError(f"Invalid day {day}. Must be 0-4")
        return sorted(set(v))


class ResidentWeeklyRequirementResponse(ResidentWeeklyRequirementBase):
    """Schema for resident weekly requirement response."""

    id: UUID
    rotation_template_id: UUID
    created_at: datetime
    updated_at: datetime

    # Computed properties
    total_min_halfdays: int = Field(description="Minimum half-days required per week")
    total_max_halfdays: int = Field(description="Maximum half-days allowed per week")
    is_valid_range: bool = Field(description="Whether min/max ranges are valid")

    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_orm_with_computed(cls, obj) -> "ResidentWeeklyRequirementResponse":
        """Create response with computed properties from ORM object."""
        return cls(
            id=obj.id,
            rotation_template_id=obj.rotation_template_id,
            fm_clinic_min_per_week=obj.fm_clinic_min_per_week,
            fm_clinic_max_per_week=obj.fm_clinic_max_per_week,
            specialty_min_per_week=obj.specialty_min_per_week,
            specialty_max_per_week=obj.specialty_max_per_week,
            academics_required=obj.academics_required,
            protected_slots=obj.protected_slots or {},
            allowed_clinic_days=obj.allowed_clinic_days or [],
            specialty_name=obj.specialty_name,
            description=obj.description,
            created_at=obj.created_at,
            updated_at=obj.updated_at,
            total_min_halfdays=obj.total_min_halfdays,
            total_max_halfdays=obj.total_max_halfdays,
            is_valid_range=obj.is_valid_range,
        )


class ResidentWeeklyRequirementListResponse(BaseModel):
    """Schema for list of resident weekly requirements."""

    items: list[ResidentWeeklyRequirementResponse]
    total: int


class ResidentWeeklyRequirementWithTemplate(ResidentWeeklyRequirementResponse):
    """Response including the associated rotation template name."""

    template_name: str = Field(description="Name of the associated rotation template")
    template_activity_type: str = Field(
        description="Activity type of the rotation template"
    )
